### PR TITLE
fix close notify

### PR DESCRIPTION
**Pull request by whyrusleeping**
Our multipart file code cannot handle sending zero files. So if the number of input files is nil, don't do any multipart stuff.

Also did some other cleanup along the way to make debugging easier.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
hrm... the test i'm writing will be a LOT easier with #2003. I'm trying to use netcat to capture the http request, but the version check goes first and i capture that instead...

**Comment by rht** [comment]
This surely fixes the closenotify problem, but not entirely obvious why.
The first roundtrip circle is somehow completed.

**Comment by rht** [comment]
#2003 is currently on dev0.4.0

**Comment by whyrusleeping** [comment]
@rht the close notify issue appeared to be because the client was sending a malformed http request body. That caused the server side to mess up for some reason and not figure out how the close notify should happen for some reason.

File: commands/http/client.go [hunk 1]
```diff
@@ -79,10 +79,6 @@ func (c *client) Send(req cmds.Request) (cmds.Response, error) {
 	if req.Files() != nil {
 		fileReader = NewMultiFileReader(req.Files(), true)
 		reader = fileReader
-	} else {
-		// if we have no file data, use an empty Reader
-		// (http.NewRequest panics when a nil Reader is used)
-		reader = strings.NewReader("")
```
Comment (jbenet) [inline commands/http/client.go]: is this no longer true, then?

File: commands/http/client.go [hunk 2]
```diff
@@ -146,8 +142,15 @@ func (c *client) Send(req cmds.Request) (cmds.Response, error) {
 }
 
 func getQuery(req cmds.Request) (string, error) {
+	// some options should be kept clientside, the server doesnt care
+	skip := map[string]bool{
+		"api": true,
+	}
```
Comment (jbenet) [inline commands/http/client.go]: :+1: should we make this a const somewhere?

File: commands/http/handler.go [hunk 3]
```diff
@@ -110,14 +109,33 @@ func (i internalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	defer cancel()
+	if cn, ok := w.(http.CloseNotifier); ok {
+		go func() {
+			select {
+			case <-cn.CloseNotify():
+			case <-ctx.Done():
+			}
+			cancel()
+		}()
+	}
```
Comment (jbenet) [inline commands/http/handler.go]: i cant comment on whether it is strictly good to do the CloseNotify before the checks below. seems right, but if others want to comment, please do

**Comment by jbenet** [comment]
this LGTM in general. 

@whyrusleeping maybe write a proxy that checks the contents are fine, but still fwds the request to something else? not sure.

File: commands/http/client.go (hunk 1 above)
Comment (whyrusleeping) [inline commands/http/client.go]: it was actually never true, as far as i can tell, at least since april 2014.

File: commands/http/client.go (hunk 2 above)
Comment (whyrusleeping) [inline commands/http/client.go]: mmm, potentially. not sure on the placement, as this function is the only spot that should be concerned with it.

**Comment by whyrusleeping** [comment]
> maybe write a proxy that checks the contents are fine, but still fwds the request to something else? not sure.

@jbenet i was going to do that with netcat, but it will be difficult before #2003 

I think I might just merge the test i have in after that one lands

**Comment by jbenet** [comment]
> @jbenet i was going to do that with netcat, but it will be difficult before #2003

that's why "proxy that checks the contents are fine, but still fwds the request to something else" -- a proxy would not have a problem with the version check, it would go through it fine.

**Comment by whyrusleeping** [comment]
I mean, if we want to have another bit of go code to do that we can. But at that point, we might as well just test it exclusively in go

**Comment by jbenet** [comment]
ok that's fine, as long as it is a real test (testing with the full api service running).

fwiw, having a tool that can proxy + verify certain things pass through will be useful anyway. imagine something like:

```sh
cat testfile
> grep ^Hello Server$
< grep ^Hello Client$
> grep ^Here are [0-9]+ things$
+> grep $Thing [0-9]+$
> grep ^Done$
< grep ^Ok got [0-9]+ things

cat testfile | testproxy --port 5679 1.2.3.4:5678
```
this may be overkill for now though :)

**Comment by rht** [comment]
For the test, #2003 -> #2037.

**Comment by cryptix** [comment]
LGTM

**Comment by whyrusleeping** [comment]
this has been long in coming and is actively causing issues, merging.

**Pull request merged**